source/sdk/lang/Exception.ooc

```diff
@@ -1,397 +1,324 @@
-        if (backtrace)
-          backtrace free()
-      }
-      this backtraces free()
-      super()
-    }
-
-    /**
-     * @return the exception's message, nicely formatted
-     */
-    formatMessage: func -> String {
-        if(origin)
-            "[%s in %s]: %s\n" format(class name toCString(), origin name toCString(), message ? message toCString() : "<no message>" toCString())
-        else
-            "[%s]: %s\n" format(class name toCString(), message ? message toCString() : "<no message>" toCString())
-    }
-
-    /**
-     * Print this exception, with its origin, if specified, and its message
-     */
-    print: func {
-        printMessage()
-        printBacktrace()
-    }
-
-    /**
-     * Print just the message
-     */
-    printMessage: func {
-        fprintf(stderr, "%s", formatMessage() toCString())
-    }
-
-    /**
-     * Throw this exception
-     */
-    throw: func {
-        _cleanupStackFrames()
-        _setException(this)
-        addBacktrace()
-        if(!_hasStackFrame()) {
-            version (windows) {
-                if (IsDebuggerPresent()) {
-                    // trigger a break point here, debugger will like that!
-                    printMessage()
-                    DebugBreak()
-                } else {
-                    // print the backtrace ourselves
-                    print()
-                }
-                exit(1)
-            }
-            version (!windows) {
-                printMessage()
-                abort()
-            }
-        } else {
-            frame := _takeStackFrame()
-            _markStackFrame(frame)
-            frame@ buf longJmp(_EXCEPTION)
-        }
-    }
-
-    /**
-     * Rethrow this exception.
-     */
-    rethrow: func {
-        throw()
-    }
-
-    getCurrentBacktrace: static func -> String {
-        result: String
-        h := BacktraceHandler get()
-        bt := h backtrace()
-        if (bt) {
-            result = h backtraceSymbols(bt)
-            bt free()
-        } else
-          result ="[no backtrace] use a debugger!\n"
-        result
-    }
+	backtraces := Stack<Backtrace> new()
+	origin: Class
+	message : String
+
+	addBacktrace: func {
+		backtrace := BacktraceHandler get() backtrace()
+		if (backtrace)
+			backtraces push(backtrace)
+	}
+	printBacktrace: func {
+		handler := BacktraceHandler get()
+		for (i in 0 .. this backtraces size)
+			stderr write(handler backtraceSymbols(this backtraces peek(i)))
+	}
+	init: func (=origin, =message)
+	init: func ~noOrigin (=message) { init(null, message) }
+	free: override func {
+		while (!this backtraces isEmpty) {
+			backtrace := this backtraces pop()
+			if (backtrace)
+				backtrace free()
+		}
+		this backtraces free()
+		super()
+	}
+	formatMessage: func -> String {
+		if (origin)
+			"[%s in %s]: %s\n" format(class name toCString(), origin name toCString(), message ? message toCString() : "<no message>" toCString())
```

Comment:
No no, it's OK:) I just really have no idea. In fact initialy I just removed the globals and cleaned up the ugly space formatting without looking at the rest of this code. As we want to keep this file as our own, let's do this cleanup properly.